// inferable/bin/index.js
"use strict";
/**
 * # inferable
 *
 * ## Installation
 *
 * ```bash
 * npm install inferable
 * ```
 *
 * ```bash
 * yarn add inferable
 * ```
 *
 * ```bash
 * pnpm add inferable
 * ```
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createApiClient = exports.blob = exports.validateFunctionSchema = exports.validateFunctionName = exports.validateServiceName = exports.validateDescription = exports.resultHandlerSchema = exports.InferablePromptfooProvider = exports.masked = exports.Inferable = void 0;
var Inferable_1 = require("./Inferable");
Object.defineProperty(exports, "Inferable", { enumerable: true, get: function () { return Inferable_1.Inferable; } });
const masked = () => {
    throw new Error("masked is not implemented");
};
exports.masked = masked;
exports.InferablePromptfooProvider = __importStar(require("./eval/promptfoo"));
var types_1 = require("./types");
Object.defineProperty(exports, "resultHandlerSchema", { enumerable: true, get: function () { return types_1.resultHandlerSchema; } });
var util_1 = require("./util");
Object.defineProperty(exports, "validateDescription", { enumerable: true, get: function () { return util_1.validateDescription; } });
Object.defineProperty(exports, "validateServiceName", { enumerable: true, get: function () { return util_1.validateServiceName; } });
Object.defineProperty(exports, "validateFunctionName", { enumerable: true, get: function () { return util_1.validateFunctionName; } });
Object.defineProperty(exports, "validateFunctionSchema", { enumerable: true, get: function () { return util_1.validateFunctionSchema; } });
Object.defineProperty(exports, "blob", { enumerable: true, get: function () { return util_1.blob; } });
var create_client_1 = require("./create-client");
Object.defineProperty(exports, "createApiClient", { enumerable: true, get: function () { return create_client_1.createApiClient; } });
