// inferable/src/contract.ts
import { initContract } from "@ts-rest/core";
import { z } from "zod";

const c = initContract();

const machineHeaders = {
  "x-machine-id": z.string().optional(),
  "x-machine-sdk-version": z.string().optional(),
  "x-machine-sdk-language": z.string().optional(),
  "x-forwarded-for": z.string().optional().optional(),
  "x-sentinel-no-mask": z.string().optional().optional(),
  "x-sentinel-unmask-keys": z.string().optional(),
};

export const blobSchema = z.object({
  id: z.string(),
  name: z.string(),
  type: z.enum(["json", "json-array"]),
  encoding: z.enum(["base64"]),
  size: z.number(),
  createdAt: z.date(),
});

export const VersionedTextsSchema = z.object({
  current: z.object({
    version: z.string(),
    content: z.string(),
  }),
  history: z.array(
    z.object({
      version: z.string(),
      content: z.string(),
    }),
  ),
});

export const genericMessageDataSchema = z
  .object({
    message: z.string(),
    details: z.object({}).passthrough().optional(),
  })
  .strict();

export const resultDataSchema = z
  .object({
    id: z.string(),
    result: z.object({}).passthrough(),
  })
  .strict();

export const learningSchema = z.object({
  summary: z
    .string()
    .describe(
      "The new information that was learned. Be generic, do not refer to the entities.",
    ),
  entities: z
    .array(
      z.object({
        name: z
          .string()
          .describe("The name of the entity this learning relates to."),
        type: z.enum(["tool"]),
      }),
    )
    .describe("The entities this learning relates to."),
  relevance: z.object({
    temporality: z
      .enum(["transient", "persistent"])
      .describe("How long do you expect this learning to be relevant for."),
  }),
});

export const agentDataSchema = z
  .object({
    done: z.boolean().optional(),
    result: z.any().optional(),
    summary: z.string().optional(),
    learnings: z.array(learningSchema).optional(),
    issue: z.string().optional(),
    invocations: z
      .array(
        z.object({
          id: z.string().optional(),
          toolName: z.string(),
          reasoning: z.string(),
          input: z.object({}).passthrough(),
        }),
      )
      .optional(),
  })
  .strict();

export const messageDataSchema = z.union([
  resultDataSchema,
  agentDataSchema,
  genericMessageDataSchema,
]);

export const FunctionConfigSchema = z.object({
  cache: z
    .object({
      keyPath: z.string(),
      ttlSeconds: z.number(),
    })
    .optional(),
  retryCountOnStall: z.number().optional(),
  timeoutSeconds: z.number().optional(),
  executionIdPath: z.string().optional(),
  requiresApproval: z.boolean().default(false).optional(),
  private: z.boolean().default(false).optional(),
});

export const definition = {
  createMachine: {
    method: "POST",
    path: "/machines",
    headers: z.object({
      authorization: z.string(),
      ...machineHeaders,
    }),
    body: z.object({
      service: z.string(),
      functions: z
        .array(
          z.object({
            name: z.string(),
            description: z.string().optional(),
            schema: z.string().optional(),
            config: FunctionConfigSchema.optional(),
          }),
        )
        .optional(),
    }),
    responses: {
      200: z.object({
        clusterId: z.string(),
        queueUrl: z.string(),
        region: z.string(),
        enabled: z.boolean().default(true),
        expiration: z.date(),
        credentials: z.object({
          accessKeyId: z.string(),
          secretAccessKey: z.string(),
          sessionToken: z.string(),
        }),
      }),
      204: z.undefined(),
    },
  },
  // TODO: Remove
  acknowledgeJob: {
    method: "PUT",
    path: "/jobs/:jobId",
    headers: z.object({
      authorization: z.string(),
      ...machineHeaders,
    }),
    pathParams: z.object({
      jobId: z.string(),
    }),
    responses: {
      204: z.undefined(),
      401: z.undefined(),
    },
    body: z.undefined(),
  },
  // TODO: Remove
  createResult: {
    method: "POST",
    path: "/jobs/:jobId/result",
    headers: z.object({
      authorization: z.string(),
      ...machineHeaders,
    }),
    pathParams: z.object({
      jobId: z.string(),
    }),
    responses: {
      204: z.undefined(),
      401: z.undefined(),
    },
    body: z.object({
      result: z.string(),
      resultType: z.enum(["resolution", "rejection"]),
      // TODO: wrap this in meta
      functionExecutionTime: z.number().optional(),
    }),
  },
  live: {
    method: "GET",
    path: "/live",
    responses: {
      200: z.object({
        status: z.string(),
      }),
    },
  },
  createBlob: {
    method: "POST",
    path: "/jobs/:jobId/blob",
    headers: z.object({
      authorization: z.string(),
      "x-machine-id": z.string(),
      "x-machine-sdk-version": z.string(),
      "x-machine-sdk-language": z.string(),
      "x-forwarded-for": z.string().optional(),
      "x-sentinel-no-mask": z.string().optional(),
    }),
    pathParams: z.object({
      jobId: z.string(),
    }),
    responses: {
      201: z.object({
        id: z.string(),
      }),
      401: z.undefined(),
      404: z.object({
        message: z.string(),
      }),
    },
    body: blobSchema.omit({ id: true, createdAt: true }).and(
      z.object({
        data: z.string(),
      }),
    ),
  },
  getContract: {
    method: "GET",
    path: "/contract",
    responses: {
      200: z.object({
        contract: z.string(),
      }),
    },
  },
  listClusters: {
    method: "GET",
    path: "/clusters",
    headers: z.object({
      authorization: z.string(),
    }),
    responses: {
      200: z.array(
        z.object({
          id: z.string(),
          name: z.string(),
          createdAt: z.date(),
          description: z.string().nullable(),
        }),
      ),
      401: z.undefined(),
    },
  },
  createCluster: {
    method: "POST",
    path: "/clusters",
    headers: z.object({
      authorization: z.string(),
    }),
    responses: {
      204: z.undefined(),
    },
    body: z.object({
      description: z
        .string()
        .describe("Human readable description of the cluster"),
    }),
  },
  updateCluster: {
    method: "PUT",
    path: "/clusters/:clusterId",
    headers: z.object({
      authorization: z.string(),
    }),
    responses: {
      204: z.undefined(),
      401: z.undefined(),
    },
    body: z.object({
      name: z.string(),
      description: z.string(),
      additionalContext: z
        .object({
          current: z
            .object({
              version: z.string(),
              content: z.string(),
            })
            .describe("Current cluster context version"),
          history: z
            .array(
              z.object({
                version: z.string(),
                content: z.string(),
              }),
            )
            .describe("History of the cluster context versions"),
        })
        .optional()
        .describe("Additional cluster context which is included in all runs"),
      debug: z
        .boolean()
        .optional()
        .describe(
          "Enable additional logging (Including prompts and results) for use by Inferable support",
        ),
    }),
  },
  getCluster: {
    method: "GET",
    path: "/clusters/:clusterId",
    headers: z.object({
      authorization: z.string(),
    }),
    responses: {
      200: z.object({
        id: z.string(),
        name: z.string(),
        description: z.string().nullable(),
        createdAt: z.date(),
        debug: z.boolean(),
        lastPingAt: z.date().nullable(),
      }),
      401: z.undefined(),
      404: z.undefined(),
    },
    pathParams: z.object({
      clusterId: z.string(),
    }),
  },
  getService: {
    method: "GET",
    path: "/clusters/:clusterId/service/:serviceName",
    headers: z.object({
      authorization: z.string(),
    }),
    responses: {
      200: z.object({
        jobs: z.array(
          z.object({
            id: z.string(),
            targetFn: z.string(),
            service: z.string().nullable(),
            status: z.string(),
            resultType: z.string().nullable(),
            createdAt: z.date(),
            functionExecutionTime: z.number().nullable(),
          }),
        ),
        definition: z
          .object({
            name: z.string(),
            functions: z
              .array(
                z.object({
                  name: z.string(),
                  rate: z
                    .object({
                      per: z.enum(["minute", "hour"]),
                      limit: z.number(),
                    })
                    .optional(),
                  cacheTTL: z.number().optional(),
                }),
              )
              .optional(),
          })
          .nullable(),
      }),
      401: z.undefined(),
      404: z.undefined(),
    },
    pathParams: z.object({
      clusterId: z.string(),
      serviceName: z.string(),
    }),
    query: z.object({
      limit: z.coerce.number().min(100).max(5000).default(2000),
    }),
  },
  listEvents: {
    method: "GET",
    path: "/clusters/:clusterId/events",
    headers: z.object({
      authorization: z.string(),
    }),
    responses: {
      200: z.array(
        z.object({
          type: z.string(),
          machineId: z.string().nullable(),
          service: z.string().nullable(),
          createdAt: z.date(),
          jobId: z.string().nullable(),
          targetFn: z.string().nullable(),
          resultType: z.string().nullable(),
          status: z.string().nullable(),
          workflowId: z.string().nullable(),
          meta: z.any().nullable(),
          id: z.string(),
        }),
      ),
      401: z.undefined(),
      404: z.undefined(),
    },
    query: z.object({
      type: z.string().optional(),
      jobId: z.string().optional(),
      machineId: z.string().optional(),
      service: z.string().optional(),
      workflowId: z.string().optional(),
      includeMeta: z.string().optional(),
    }),
  },
  getEventMeta: {
    method: "GET",
    path: "/clusters/:clusterId/events/:eventId/meta",
    headers: z.object({
      authorization: z.string(),
    }),
    responses: {
      200: z.object({
        type: z.string(),
        machineId: z.string().nullable(),
        service: z.string().nullable(),
        createdAt: z.date(),
        jobId: z.string().nullable(),
        targetFn: z.string().nullable(),
        resultType: z.string().nullable(),
        status: z.string().nullable(),
        meta: z.unknown(),
        id: z.string(),
      }),
      401: z.undefined(),
      404: z.undefined(),
    },
  },
  // TODO: Remove
  executeJobSync: {
    method: "POST",
    path: "/clusters/:clusterId/execute",
    headers: z.object({
      authorization: z.string(),
    }),
    body: z.object({
      service: z.string(),
      function: z.string(),
      input: z.object({}).passthrough(),
    }),
    responses: {
      401: z.undefined(),
      404: z.undefined(),
      200: z.object({
        resultType: z.string(),
        result: z.any(),
        status: z.string(),
      }),
      400: z.object({
        message: z.string(),
      }),
      500: z.object({
        error: z.string(),
      }),
    },
  },
  createRun: {
    method: "POST",
    path: "/clusters/:clusterId/runs",
    headers: z.object({
      authorization: z.string(),
    }),
    body: z.object({
      message: z
        .string()
        .optional()
        .describe(
          "The prompt message, do not provide if using a prompt template",
        ),
      result: z
        .object({
          handler: z
            .object({
              service: z.string(),
              function: z.string(),
            })
            .optional()
            .describe(
              "The Inferable function which will be used to return result for the run",
            ),
          schema: z
            .object({})
            .passthrough()
            .optional()
            .describe("The JSON schema which the result should conform to"),
        })
        .optional(),
      attachedFunctions: z
        .array(z.string())
        .optional()
        .describe(
          "An array of attached functions (Keys should be service in the format <SERVICE>_<FUNCTION>)",
        ),
      metadata: z
        .record(z.string())
        .optional()
        .describe("Run metadata which can be used to filter runs"),
      test: z
        .object({
          enabled: z.boolean().default(false),
          mocks: z
            .record(
              z.object({
                output: z
                  .object({})
                  .passthrough()
                  .describe("The mock output of the function"),
              }),
            )
            .optional()
            .describe(
              "Function mocks to be used in the run. (Keys should be function in the format <SERVICE>_<FUNCTION>)",
            ),
        })
        .optional()
        .describe(
          "When provided, the run will be marked as as a test / evaluation",
        ),
      template: z
        .object({
          id: z.string().describe("The prompt template ID"),
          input: z
            .object({})
            .passthrough()
            .describe(
              "The input arguments, these should match what is described in the prompt template definition",
            ),
        })
        .optional()
        .describe("A prompt template which the run should be created from"),
    }),
    responses: {
      201: z.object({
        id: z.string().describe("The id of the newly created run"),
      }),
      401: z.undefined(),
      400: z.object({
        message: z.string(),
      }),
    },
    pathParams: z.object({
      clusterId: z.string(),
    }),
  },
  deleteRun: {
    method: "DELETE",
    path: "/clusters/:clusterId/runs/:runId",
    headers: z.object({
      authorization: z.string(),
    }),
    body: z.undefined(),
    responses: {
      204: z.undefined(),
      401: z.undefined(),
    },
    pathParams: z.object({
      runId: z.string(),
      clusterId: z.string(),
    }),
  },
  createMessage: {
    method: "POST",
    path: "/clusters/:clusterId/runs/:runId/messages",
    headers: z.object({
      authorization: z.string(),
    }),
    body: z.object({
      id: z
        .string()
        .length(26)
        .regex(/^[0-9a-z]+$/i)
        .optional(),
      message: z.string(),
      type: z.enum(["human", "supervisor"]).optional(),
    }),
    responses: {
      200: z.object({
        messages: z.array(
          z.object({
            id: z.string(),
            data: messageDataSchema,
            type: z.enum([
              "human",
              "template",
              "result",
              "agent",
              "agent-invalid",
              "supervisor",
            ]),
            createdAt: z.date(),
            pending: z.boolean().default(false),
          }),
        ),
      }),
      401: z.undefined(),
    },
    pathParams: z.object({
      runId: z.string(),
      clusterId: z.string(),
    }),
  },
  listMessages: {
    method: "GET",
    path: "/clusters/:clusterId/runs/:runId/messages",
    headers: z.object({
      authorization: z.string(),
    }),
    responses: {
      200: z.array(
        z.object({
          id: z.string(),
          data: messageDataSchema,
          type: z.enum([
            "human",
            "template",
            "result",
            "agent",
            "agent-invalid",
            "supervisor",
          ]),
          createdAt: z.date(),
          pending: z.boolean().default(false),
          displayableContext: z.record(z.string()).nullable(),
        }),
      ),
      401: z.undefined(),
    },
  },
  listRuns: {
    method: "GET",
    path: "/clusters/:clusterId/runs",
    headers: z.object({
      authorization: z.string(),
    }),
    query: z.object({
      userId: z.string().optional(),
      test: z.coerce
        .string()
        .transform((value) => value === "true")
        .optional(),
      limit: z.coerce.number().min(10).max(50).default(50),
      metadata: z
        .string()
        .optional()
        .describe("Filter runs by a metadata value (value:key)"),
      promptTemplateId: z.string().optional(),
    }),
    responses: {
      200: z.array(
        z.object({
          id: z.string(),
          name: z.string(),
          userId: z.string().nullable(),
          createdAt: z.date(),
          status: z
            .enum(["pending", "running", "paused", "done", "failed"])
            .nullable(),
          parentWorkflowId: z.string().nullable(),
          test: z.boolean(),
          promptTemplateId: z.string().nullable(),
          promptTemplateVersion: z.number().nullable(),
          feedbackScore: z.number().nullable(),
        }),
      ),
      401: z.undefined(),
    },
  },
  getRun: {
    method: "GET",
    path: "/clusters/:clusterId/runs/:runId",
    headers: z.object({
      authorization: z.string(),
    }),
    responses: {
      200: z.object({
        id: z.string(),
        jobHandle: z.string().nullable(),
        userId: z.string().nullable(),
        status: z
          .enum(["pending", "running", "paused", "done", "failed"])
          .nullable(),
        failureReason: z.string().nullable(),
        test: z.boolean(),
        feedbackComment: z.string().nullable(),
        feedbackScore: z.number().nullable(),
        result: z.string().nullable(),
        summary: z.string().nullable(),
        metadata: z.record(z.string()).nullable(),
        attachedFunctions: z.array(z.string()).nullable(),
      }),
      401: z.undefined(),
    },
  },
  createFeedback: {
    method: "POST",
    path: "/clusters/:clusterId/runs/:runId/feedback",
    headers: z.object({
      authorization: z.string(),
    }),
    body: z.object({
      comment: z.string().describe("Feedback comment").nullable(),
      score: z
        .number()
        .describe("Score between 0 and 1")
        .min(0)
        .max(1)
        .nullable(),
    }),
    responses: {
      204: z.undefined(),
      401: z.undefined(),
      404: z.undefined(),
    },
    pathParams: z.object({
      runId: z.string(),
      clusterId: z.string(),
    }),
  },
  resolveInputRequest: {
    method: "POST",
    path: "/clusters/:clusterId/runs/:runId/input-requests/:inputRequestId",
    headers: z.object({
      authorization: z.string().optional(),
    }),
    body: z.object({
      input: z.string(),
    }),
    responses: {
      204: z.undefined(),
      401: z.undefined(),
      404: z.undefined(),
    },
    pathParams: z.object({
      runId: z.string(),
      inputRequestId: z.string(),
      clusterId: z.string(),
    }),
  },
  getInputRequest: {
    method: "GET",
    path: "/clusters/:clusterId/runs/:runId/input-requests/:inputRequestId",
    headers: z.object({
      authorization: z.string().optional(),
    }),
    pathParams: z.object({
      clusterId: z.string(),
      runId: z.string(),
      inputRequestId: z.string(),
    }),
    responses: {
      200: z.object({
        id: z.string(),
        runId: z.string(),
        clusterId: z.string(),
        resolvedAt: z.date().nullable(),
        createdAt: z.date(),
        requestArgs: z.string().nullable(),
        service: z.string().nullable(),
        function: z.string().nullable(),
        description: z.string().nullable(),
        type: z.string(),
        options: z.array(z.string()).optional(),
      }),
      401: z.undefined(),
      404: z.undefined(),
    },
  },
  oas: {
    method: "GET",
    path: "/public/oas.json",
    responses: {
      200: z.unknown(),
    },
  },
  // TODO: Remove
  pingCluster: {
    method: "POST",
    path: "/ping-cluster",
    headers: z.object({
      authorization: z.string(),
      "x-machine-id": z.string(),
      "x-machine-sdk-version": z.string(),
      "x-machine-sdk-language": z.string(),
      "x-forwarded-for": z.string().optional(),
    }),
    body: z.object({
      services: z.array(z.string()),
    }),
    responses: {
      204: z.undefined(),
      401: z.undefined(),
    },
  },
  // TODO: Remove
  pingClusterV2: {
    method: "POST",
    path: "/ping-cluster-v2",
    headers: z.object({
      authorization: z.string(),
      "x-machine-id": z.string(),
      "x-machine-sdk-version": z.string(),
      "x-machine-sdk-language": z.string(),
      "x-forwarded-for": z.string().optional(),
      "x-sentinel-no-mask": z.string().optional(),
    }),
    body: z.object({
      services: z.array(z.string()),
    }),
    responses: {
      200: z.object({
        outdated: z.boolean(),
      }),
      401: z.undefined(),
    },
  },
  updateMessage: {
    method: "PUT",
    path: "/clusters/:clusterId/runs/:runId/messages/:messageId",
    headers: z.object({ authorization: z.string() }),
    body: z.object({ message: z.string() }),
    responses: {
      200: z.object({
        data: genericMessageDataSchema,
        id: z.string(),
      }),
      404: z.object({ message: z.string() }),
      401: z.undefined(),
    },
    pathParams: z.object({
      clusterId: z.string(),
      runId: z.string(),
      messageId: z.string(),
    }),
  },
  storeServiceMetadata: {
    method: "PUT",
    path: "/clusters/:clusterId/services/:service/keys/:key",
    headers: z.object({ authorization: z.string() }),
    body: z.object({
      value: z.string(),
    }),
    pathParams: z.object({
      clusterId: z.string(),
      service: z.string(),
      key: z.string(),
    }),
    responses: {
      204: z.undefined(),
      401: z.undefined(),
    },
  },
  getClusterExport: {
    method: "GET",
    path: "/clusters/:clusterId/export",
    headers: z.object({ authorization: z.string() }),
    pathParams: z.object({ clusterId: z.string() }),
    responses: {
      200: z.object({
        data: z.string(),
      }),
    },
  },
  consumeClusterExport: {
    method: "POST",
    path: "/clusters/:clusterId/import",
    headers: z.object({ authorization: z.string() }),
    body: z.object({ data: z.string() }),
    pathParams: z.object({ clusterId: z.string() }),
    responses: {
      200: z.object({ message: z.string() }),
      400: z.undefined(),
    },
  },
  // TODO: Remove
  getJob: {
    method: "GET",
    path: "/clusters/:clusterId/jobs/:jobId",
    headers: z.object({ authorization: z.string() }),
    pathParams: z.object({
      clusterId: z.string(),
      jobId: z.string(),
    }),
    responses: {
      200: z.object({
        id: z.string(),
        status: z.string(),
        targetFn: z.string(),
        service: z.string(),
        executingMachineId: z.string().nullable(),
        targetArgs: z.string(),
        result: z.string().nullable(),
        resultType: z.string().nullable(),
        createdAt: z.date(),
        blobs: z.array(blobSchema),
      }),
    },
  },
  listJobReferences: {
    method: "GET",
    path: "/clusters/:clusterId/runs/:runId/job-references",
    headers: z.object({ authorization: z.string() }),
    pathParams: z.object({
      clusterId: z.string(),
      runId: z.string(),
    }),
    query: z.object({
      token: z.string(),
      before: z.string(),
    }),
    responses: {
      200: z.array(
        z.object({
          id: z.string(),
          result: z.string().nullable(),
          createdAt: z.date(),
          status: z.string(),
          targetFn: z.string(),
          service: z.string(),
          executingMachineId: z.string().nullable(),
        }),
      ),
    },
  },
  createApiKey: {
    method: "POST",
    path: "/clusters/:clusterId/api-keys",
    headers: z.object({ authorization: z.string() }),
    pathParams: z.object({
      clusterId: z.string(),
    }),
    body: z.object({
      name: z.string(),
      type: z.enum(["cluster_manage", "cluster_consume", "cluster_machine"]),
    }),
    responses: {
      200: z.object({
        id: z.string(),
        key: z.string(),
      }),
    },
  },
  listApiKeys: {
    method: "GET",
    path: "/clusters/:clusterId/api-keys",
    headers: z.object({ authorization: z.string() }),
    pathParams: z.object({
      clusterId: z.string(),
    }),
    responses: {
      200: z.array(
        z.object({
          id: z.string(),
          name: z.string(),
          type: z.enum([
            "cluster_manage",
            "cluster_consume",
            "cluster_machine",
          ]),
          createdAt: z.date(),
          createdBy: z.string(),
          revokedAt: z.date().nullable(),
        }),
      ),
    },
  },
  revokeApiKey: {
    method: "DELETE",
    path: "/clusters/:clusterId/api-keys/:keyId",
    headers: z.object({ authorization: z.string() }),
    pathParams: z.object({
      clusterId: z.string(),
      keyId: z.string(),
    }),
    body: z.undefined(),
    responses: {
      204: z.undefined(),
    },
  },
  getClusterContext: {
    method: "GET",
    path: "/clusters/:clusterId/additional-context",
    headers: z.object({
      authorization: z.string(),
    }),
    responses: {
      200: z.object({
        additionalContext: VersionedTextsSchema.nullable(),
      }),
      401: z.undefined(),
      404: z.undefined(),
    },
    pathParams: z.object({
      clusterId: z.string(),
    }),
  },
  listMachines: {
    method: "GET",
    path: "/clusters/:clusterId/machines",
    headers: z.object({
      authorization: z.string(),
    }),
    responses: {
      200: z.array(
        z.object({
          id: z.string(),
          lastPingAt: z.date(),
          ip: z.string(),
        }),
      ),
    },
    pathParams: z.object({
      clusterId: z.string(),
    }),
  },
  listServices: {
    method: "GET",
    path: "/clusters/:clusterId/services",
    headers: z.object({
      authorization: z.string(),
    }),
    responses: {
      200: z.array(
        z.object({
          name: z.string(),
          description: z.string().optional(),
          functions: z
            .array(
              z.object({
                name: z.string(),
                description: z.string().optional(),
                schema: z.string().optional(),
                config: FunctionConfigSchema.optional(),
              }),
            )
            .optional(),
        }),
      ),
    },
    pathParams: z.object({
      clusterId: z.string(),
    }),
  },
  getRunTimeline: {
    method: "GET",
    path: "/clusters/:clusterId/runs/:runId/timeline",
    headers: z.object({ authorization: z.string() }),
    query: z.object({
      messagesAfter: z.string().default("0"),
      activityAfter: z.string().default("0"),
      jobsAfter: z.string().default("0"),
      inputRequestsAfter: z.string().default("0"),
    }),
    pathParams: z.object({
      clusterId: z.string(),
      runId: z.string(),
    }),
    responses: {
      200: z.object({
        messages: z.array(
          z.object({
            id: z.string(),
            data: messageDataSchema,
            type: z.enum([
              "human",
              // TODO: Remove 'template' type
              "template",
              "result",
              "agent",
              "agent-invalid",
              "supervisor",
            ]),
            createdAt: z.date(),
            pending: z.boolean().default(false),
            displayableContext: z.record(z.string()).nullable(),
          }),
        ),
        activity: z.array(
          z.object({
            id: z.string(),
            type: z.string(),
            machineId: z.string().nullable(),
            service: z.string().nullable(),
            createdAt: z.date(),
            jobId: z.string().nullable(),
            targetFn: z.string().nullable(),
          }),
        ),
        jobs: z.array(
          z.object({
            id: z.string(),
            status: z.string(),
            targetFn: z.string(),
            service: z.string(),
            resultType: z.string().nullable(),
            createdAt: z.date(),
          }),
        ),
        inputRequests: z.array(
          z.object({
            id: z.string(),
            type: z.string(),
            requestArgs: z.string().nullable().optional(),
            resolvedAt: z.date().nullable().optional(),
            createdAt: z.date(),
            service: z.string().nullable().optional(),
            function: z.string().nullable().optional(),
            description: z.string().nullable().optional(),
            presentedOptions: z.array(z.string()).nullable().optional(),
          }),
        ),
        run: z.object({
          id: z.string(),
          jobHandle: z.string().nullable(),
          userId: z.string().nullable(),
          status: z
            .enum(["pending", "running", "paused", "done", "failed"])
            .nullable(),
          failureReason: z.string().nullable(),
          test: z.boolean(),
          feedbackComment: z.string().nullable(),
          feedbackScore: z.number().nullable(),
          attachedFunctions: z.array(z.string()).nullable(),
          name: z.string().nullable(),
        }),
        blobs: z.array(blobSchema),
      }),
    },
  },
  getBlobData: {
    method: "GET",
    path: "/clusters/:clusterId/blobs/:blobId/data",
    headers: z.object({ authorization: z.string() }),
    pathParams: z.object({
      clusterId: z.string(),
      blobId: z.string(),
    }),
    responses: {
      200: z.any(),
      404: z.undefined(),
    },
  },
  upsertToolMetadata: {
    method: "PUT",
    path: "/clusters/:clusterId/tool-metadata",
    headers: z.object({
      authorization: z.string(),
    }),
    pathParams: z.object({
      clusterId: z.string(),
    }),
    body: z.object({
      service: z.string(),
      function_name: z.string(),
      user_defined_context: z.string().nullable(),
      result_schema: z.unknown().nullable(),
    }),
    responses: {
      204: z.undefined(),
      401: z.undefined(),
    },
  },
  getToolMetadata: {
    method: "GET",
    path: "/clusters/:clusterId/tool-metadata/:service/:function_name",
    headers: z.object({
      authorization: z.string(),
    }),
    pathParams: z.object({
      clusterId: z.string(),
      service: z.string(),
      function_name: z.string(),
    }),
    responses: {
      200: z.object({
        cluster_id: z.string(),
        service: z.string(),
        function_name: z.string(),
        user_defined_context: z.string().nullable(),
        result_schema: z.unknown().nullable(),
      }),
      401: z.undefined(),
      404: z.object({ message: z.string() }),
    },
  },
  getAllToolMetadataForService: {
    method: "GET",
    path: "/clusters/:clusterId/tool-metadata/:service",
    headers: z.object({
      authorization: z.string(),
    }),
    pathParams: z.object({
      clusterId: z.string(),
      service: z.string(),
    }),
    responses: {
      200: z.array(
        z.object({
          cluster_id: z.string(),
          service: z.string(),
          function_name: z.string(),
          user_defined_context: z.string().nullable(),
          result_schema: z.unknown().nullable(),
        }),
      ),
      401: z.undefined(),
    },
  },
  deleteToolMetadata: {
    method: "DELETE",
    path: "/clusters/:clusterId/tool-metadata/:service/:function_name",
    headers: z.object({
      authorization: z.string(),
    }),
    pathParams: z.object({
      clusterId: z.string(),
      service: z.string(),
      function_name: z.string(),
    }),
    body: z.undefined(),
    responses: {
      204: z.undefined(),
      401: z.undefined(),
    },
  },
  generatePromptTemplate: {
    method: "GET",
    path: "/clusters/:clusterId/prompt-templates/generate",
    headers: z.object({ authorization: z.string() }),
    query: z.object({
      runId: z.string(),
      messageId: z.string().optional(),
    }),
    responses: {
      200: z.object({
        name: z.string(),
        prompt: z.string(),
        attachedFunctions: z.array(z.string()),
        structuredOutput: z.unknown().nullable(),
      }),
    },
  },
  createPromptTemplate: {
    method: "POST",
    path: "/clusters/:clusterId/prompt-templates",
    headers: z.object({ authorization: z.string() }),
    body: z.object({
      name: z.string(),
      prompt: z.string(),
      attachedFunctions: z.array(z.string()),
      structuredOutput: z.object({}).passthrough().optional(),
    }),
    responses: {
      201: z.object({ id: z.string() }),
      401: z.undefined(),
    },
    pathParams: z.object({
      clusterId: z.string(),
    }),
  },
  getPromptTemplate: {
    method: "GET",
    path: "/clusters/:clusterId/prompt-templates/:templateId",
    headers: z.object({ authorization: z.string() }),
    responses: {
      200: z.object({
        id: z.string(),
        clusterId: z.string(),
        name: z.string(),
        prompt: z.string(),
        attachedFunctions: z.array(z.string()),
        structuredOutput: z.unknown().nullable(),
        createdAt: z.date(),
        updatedAt: z.date(),
        versions: z.array(
          z.object({
            version: z.number(),
            name: z.string(),
            prompt: z.string(),
            attachedFunctions: z.array(z.string()),
            structuredOutput: z.unknown().nullable(),
          }),
        ),
      }),
      401: z.undefined(),
      404: z.object({ message: z.string() }),
    },
    pathParams: z.object({
      clusterId: z.string(),
      templateId: z.string(),
    }),
    query: z.object({
      withPreviousVersions: z.enum(["true", "false"]).default("false"),
    }),
  },
  updatePromptTemplate: {
    method: "PUT",
    path: "/clusters/:clusterId/prompt-templates/:templateId",
    headers: z.object({ authorization: z.string() }),
    body: z.object({
      name: z.string().optional(),
      prompt: z.string().optional(),
      attachedFunctions: z.array(z.string()).optional(),
      structuredOutput: z.object({}).passthrough().optional(),
    }),
    responses: {
      200: z.object({
        id: z.string(),
        clusterId: z.string(),
        name: z.string(),
        prompt: z.string(),
        attachedFunctions: z.array(z.string()),
        structuredOutput: z.unknown().nullable(),
        createdAt: z.date(),
        updatedAt: z.date(),
      }),
      401: z.undefined(),
      404: z.object({ message: z.string() }),
    },
    pathParams: z.object({
      clusterId: z.string(),
      templateId: z.string(),
    }),
  },
  deletePromptTemplate: {
    method: "DELETE",
    path: "/clusters/:clusterId/prompt-templates/:templateId",
    headers: z.object({ authorization: z.string() }),
    responses: {
      204: z.undefined(),
      401: z.undefined(),
      404: z.object({ message: z.string() }),
    },
    body: z.undefined(),
    pathParams: z.object({
      clusterId: z.string(),
      templateId: z.string(),
    }),
  },
  listPromptTemplates: {
    method: "GET",
    path: "/clusters/:clusterId/prompt-templates",
    headers: z.object({ authorization: z.string() }),
    responses: {
      200: z.array(
        z.object({
          id: z.string(),
          clusterId: z.string(),
          name: z.string(),
          prompt: z.string(),
          attachedFunctions: z.array(z.string()),
          structuredOutput: z.unknown().nullable(),
          createdAt: z.date(),
          updatedAt: z.date(),
        }),
      ),
      401: z.undefined(),
    },
    pathParams: z.object({
      clusterId: z.string(),
    }),
  },
  searchPromptTemplates: {
    method: "GET",
    path: "/clusters/:clusterId/prompt-templates/search",
    headers: z.object({ authorization: z.string() }),
    query: z.object({
      search: z.string(),
    }),
    responses: {
      200: z.array(
        z.object({
          id: z.string(),
          clusterId: z.string(),
          name: z.string(),
          prompt: z.string(),
          attachedFunctions: z.array(z.string()),
          structuredOutput: z.unknown().nullable(),
          createdAt: z.date(),
          updatedAt: z.date(),
          similarity: z.number(),
        }),
      ),
      401: z.undefined(),
    },
    pathParams: z.object({
      clusterId: z.string(),
    }),
  },
  updateClusterContext: {
    method: "PUT",
    path: "/clusters/:clusterId/additional-context",
    headers: z.object({ authorization: z.string() }),
    body: z.object({
      additionalContext: VersionedTextsSchema,
    }),
    responses: {
      204: z.undefined(),
      401: z.undefined(),
    },
    pathParams: z.object({
      clusterId: z.string(),
    }),
  },
  getTemplateMetrics: {
    method: "GET",
    path: "/clusters/:clusterId/prompt-templates/:templateId/metrics",
    headers: z.object({ authorization: z.string() }),
    responses: {
      200: z.array(
        z.object({
          createdAt: z.date(),
          feedbackScore: z.number().nullable(),
          jobFailureCount: z.number(),
          timeToCompletion: z.number(),
          jobCount: z.number(),
        }),
      ),
    },
    pathParams: z.object({
      clusterId: z.string(),
      templateId: z.string(),
    }),
  },
  createClusterKnowledgeArtifact: {
    method: "POST",
    path: "/clusters/:clusterId/knowledge",
    headers: z.object({ authorization: z.string() }),
    body: z.object({
      artifacts: z.array(
        z.object({
          id: z.string(),
          data: z.string(),
        }),
      ),
    }),
    responses: {
      204: z.undefined(),
      401: z.undefined(),
    },
    pathParams: z.object({
      clusterId: z.string(),
    }),
  },
  createRetry: {
    method: "POST",
    path: "/clusters/:clusterId/runs/:runId/retry",
    headers: z.object({ authorization: z.string() }),
    body: z.object({
      messageId: z.string(),
    }),
    responses: {
      204: z.undefined(),
      401: z.undefined(),
    },
  },
  createCall: {
    method: "POST",
    path: "/clusters/:clusterId/calls",
    query: z.object({
      waitTime: z.coerce
        .number()
        .min(0)
        .max(20)
        .default(0)
        .describe(
          "Time in seconds to keep the request open waiting for a response",
        ),
    }),
    headers: z.object({
      authorization: z.string(),
    }),
    body: z.object({
      service: z.string(),
      function: z.string(),
      input: z.object({}).passthrough(),
    }),
    responses: {
      401: z.undefined(),
      200: z.object({
        id: z.string(),
        result: z.any().nullable(),
        resultType: z.enum(["resolution", "rejection"]).nullable(),
        status: z.enum(["pending", "running", "success", "failure", "stalled"]),
      }),
    },
  },
  createCallResult: {
    method: "POST",
    path: "/clusters/:clusterId/calls/:callId/result",
    headers: z.object({
      authorization: z.string(),
      ...machineHeaders,
    }),
    pathParams: z.object({
      clusterId: z.string(),
      callId: z.string(),
    }),
    responses: {
      204: z.undefined(),
      401: z.undefined(),
    },
    body: z.object({
      result: z.any(),
      resultType: z.enum(["resolution", "rejection"]),
      meta: z.object({
        functionExecutionTime: z.number().optional(),
      }),
    }),
  },
  listCalls: {
    method: "GET",
    path: "/clusters/:clusterId/calls",
    query: z.object({
      service: z.string(),
      status: z
        .enum(["pending", "running", "paused", "done", "failed"])
        .default("pending"),
      limit: z.coerce.number().min(1).max(20).default(10),
      acknowledge: z.coerce
        .boolean()
        .default(false)
        .describe("Should calls be marked as running"),
    }),
    pathParams: z.object({
      clusterId: z.string(),
    }),
    headers: z.object({
      authorization: z.string(),
      ...machineHeaders,
    }),
    responses: {
      401: z.undefined(),
      200: z.array(
        z.object({
          id: z.string(),
          function: z.string(),
          input: z.any(),
        }),
      ),
    },
  },
} as const;

export const contract = c.router(definition);
