// inferable/src/service.ts
import debug from "debug";
import { z } from "zod";
import { createApiClient } from "./create-client";
import { InferableError } from "./errors";
import { serializeError } from "./serialize-error";
import { executeFn, Result } from "./execute-fn";
import { FunctionRegistration } from "./types";
import { extractBlobs } from "./util";

const MAX_CONSECUTIVE_POLL_FAILURES = 50;
const DEFAULT_RETRY_AFTER_SECONDS = 10;

export const log = debug("inferable:client:polling-agent");

type CallMessage = {
  id: string;
  function: string;
  input?: unknown;
};

export class Service {
  public name: string;
  public clusterId: string | null = null;
  public polling = false;

  private functions: FunctionRegistration[] = [];

  private client: ReturnType<typeof createApiClient>;

  private retryAfter = DEFAULT_RETRY_AFTER_SECONDS;

  constructor(options: {
    endpoint: string;
    machineId: string;
    apiSecret: string;
    service: string;
    functions: FunctionRegistration[];
  }) {
    this.name = options.service;

    this.client = createApiClient({
      baseUrl: options.endpoint,
      machineId: options.machineId,
      apiSecret: options.apiSecret,
    });

    this.functions = options.functions;
  }

  public async start(): Promise<{ clusterId: string }> {
    log("Starting polling agent", { service: this.name });
    const { clusterId } = await this.registerMachine();

    this.clusterId = clusterId;

    this.polling = true;

    // Purposefully not awaited
    this.runLoop();

    return {
      clusterId,
    };
  }

  public async stop(): Promise<void> {
    log("Stopping polling agent", { service: this.name });
    this.polling = false;
  }

  private async registerMachine(): Promise<{
    clusterId: string;
  }> {
    log("registering machine", {
      service: this.name,
      functions: this.functions.map((f) => f.name),
    });

    const registerResult = await this.client.createMachine({
      headers: {
        "x-sentinel-no-mask": "1",
      },
      body: {
        service: this.name,
        functions: this.functions.map((func) => ({
          name: func.name,
          description: func.description,
          schema: func.schema.inputJson,
          config: func.config,
        })),
      },
    });

    if (registerResult?.status !== 200) {
      log("Failed to register machine", registerResult);

      throw new InferableError("Failed to register machine", {
        status: registerResult.status,
        body: registerResult.body,
      });
    }

    return {
      clusterId: registerResult.body.clusterId,
    };
  }

  private async runLoop() {
    let failureCount = 0;
    while (this.polling && failureCount < MAX_CONSECUTIVE_POLL_FAILURES) {
      try {
        await this.pollIteration();
      } catch (e) {
        log("Failed poll iteration", e);
        failureCount++;
      }
      await new Promise((resolve) =>
        setTimeout(resolve, this.retryAfter * 1000),
      );
    }

    log("Quitting polling agent", { service: this.name, failureCount });
  }

  private async pollIteration() {
    if (!this.clusterId) {
      throw new Error("Failed to poll. Could not find clusterId");
    }

    const pollResult = await this.client.listCalls({
      params: {
        clusterId: this.clusterId,
      },
      query: {
        service: this.name,
        status: "pending",
        acknowledge: true,
        limit: 10,
      },
    });

    const retryAfterHeader = pollResult.headers.get("retry-after");
    if (retryAfterHeader && !isNaN(Number(retryAfterHeader))) {
      this.retryAfter = Number(retryAfterHeader);
    }

    if (pollResult?.status !== 200) {
      throw new InferableError("Failed to fetch calls", {
        status: pollResult?.status,
        body: pollResult?.body,
      });
    }

    await Promise.all(
      pollResult.body.map(async (job) => {
        await this.processCall(job);
      }),
    );
  }

  private async processCall(call: CallMessage): Promise<void> {
    const registration = this.functions.find((fn) => fn.name === call.function);

    if (!registration) {
      log("Received call for unknown function", {
        service: this.name,
        function: call.function,
      });
      return;
    }

    log("Executing job", {
      id: call.id,
      function: call.function,
      registered: !!registration,
    });

    const onComplete = async (result: Result) => {
      log("Persisting job result", {
        id: call.id,
        function: call.function,
        resultType: result.type,
        functionExecutionTime: result.functionExecutionTime,
      });

      const contentAndBlobs = extractBlobs(result.content);

      const persistResult = this.client
        .createCallResult({
          headers: {
            "x-sentinel-unmask-keys": "resultType,functionExecutionTime",
          },
          body: {
            result: contentAndBlobs.content,
            resultType: result.type,
            meta: {
              functionExecutionTime: result.functionExecutionTime,
            },
          },
          params: {
            callId: call.id,
            clusterId: this.clusterId!,
          },
        })
        .then(async (res) => {
          if (res.status === 204) {
            log("Completed job", call.id, call.function);
          } else {
            throw new InferableError(`Failed to persist call: ${res.status}`, {
              jobId: call.id,
              body: res.body,
            });
          }
        });

      const persistBlobs = contentAndBlobs.blobs.map((blob) =>
        this.client.createBlob({
          headers: {
            "x-sentinel-no-mask": "1",
          },
          params: {
            jobId: call.id,
          },
          body: blob,
        }),
      );

      await Promise.all([persistResult, ...persistBlobs]);
    };

    const args = call.input;

    log("Executing fn", {
      id: call.id,
      function: call.function,
      registeredFn: registration.func,
      args,
    });

    if (typeof args !== "object" || Array.isArray(args) || args === null) {
      log(
        "Function was called with invalid invalid format. Expected an object.",
        {
          function: call.function,
          service: this.name,
        },
      );

      return onComplete({
        type: "rejection",
        content: serializeError(
          new Error(
            "Function was called with invalid invalid format. Expected an object.",
          ),
        ),
        functionExecutionTime: 0,
      });
    }

    try {
      registration.schema.input.parse(args);
    } catch (e: unknown) {
      if (e instanceof z.ZodError) {
        e.errors.forEach((error) => {
          log("Function input does not match schema", {
            function: call.function,
            path: error.path,
            error: error.message,
          });
        });
      }

      return onComplete({
        type: "rejection",
        content: serializeError(e),
        functionExecutionTime: 0,
      });
    }

    const result = await executeFn(
      registration.func,
      [args],
      registration.authenticate,
    );

    await onComplete(result);
  }
}
